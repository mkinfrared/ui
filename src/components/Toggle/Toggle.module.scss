@import "../../styles/common";

.Toggle {
  --color: var(--toggle-text-color, var(--white));
  --padding: 0.2rem;
  --size: var(--text-height);

  align-items: center;
  display: flex;
  width: fit-content;

  .container {
    align-items: center;
    border-radius: 0.4rem;
    display: flex;
  }

  &.error {
    .switch {
      background: var(--error-gradient);
    }

    .text {
      color: var(--error);
    }
  }

  .switch {
    align-items: center;
    background: var(--secondary-gradient);
    border-radius: 5rem;
    display: flex;
    height: var(--size);
    padding: var(--padding);
    transition: all 0.25s ease-in-out;
    width: calc(var(--size) * 2 - var(--padding) * 2);

    .ball {
      aspect-ratio: 1/1;
      background-color: var(--white);
      border-radius: 50%;
      box-shadow: 0 0 0.5rem rgb(0 0 0 / 30%);
      display: block;
      height: 100%;
      transition: all 0.25s;
    }
  }

  .text {
    margin-left: 1rem;
  }

  .input {
    height: 0;
    margin: 0;
    opacity: 0;
    position: absolute;
    width: 0;
    z-index: -1;

    &:checked + .container .switch {
      background: var(--primary-gradient);
    }

    &:checked + .container .switch .ball {
      transform: translateX(100%);
    }

    &:focus ~ .container {
      border-radius: 0.4rem;
    }
  }

  input:focus-visible ~ .container {
    @include focus-visible;
  }

  &.readOnly {
    pointer-events: none;
  }

  &.readOnly .input,
  .input:disabled {
    ~ .container {
      @include disabled;

      .text {
        color: var(--grey);
      }
    }
  }
}
