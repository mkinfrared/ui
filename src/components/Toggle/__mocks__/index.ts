export { default } from "./Toggle.mock";
