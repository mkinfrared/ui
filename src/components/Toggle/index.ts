export { default } from "./Toggle";

export * from "./Toggle.type";
