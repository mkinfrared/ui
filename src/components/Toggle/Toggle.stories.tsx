import { Meta, Story } from "@storybook/react";
import React, { useState } from "react";

import css from "./Story.module.scss";
import { Toggle } from "./Toggle";
import { ToggleProps } from "./Toggle.type";

export default {
  title: "UI/Toggle",
  component: Toggle,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const UncontrolledTemplate: Story<ToggleProps> = (args) => (
  <div className={css.Story}>
    <Toggle {...args} />
  </div>
);

const ControlledTemplate: Story<ToggleProps> = () => {
  const [checked, setChecked] = useState(false);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { target } = event;

    setChecked(target.checked);
  };

  return (
    <div className={css.Story}>
      <Toggle name="example" onChange={handleChange} checked={checked} />
    </div>
  );
};

const Uncontrolled = UncontrolledTemplate.bind({});
const Controlled = ControlledTemplate.bind({});

Uncontrolled.args = {
  label: "marklar",
};

Uncontrolled.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Uncontrolled, Controlled };
