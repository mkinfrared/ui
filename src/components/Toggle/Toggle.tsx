import React, { memo } from "react";

import Text from "components/Text";
import { classNames, mergeRefs } from "utils";

import css from "./Toggle.module.scss";
import { ToggleProps } from "./Toggle.type";

/**
 * renders a an iOS like toggle
 */
const Toggle = ({
  checked,
  className,
  defaultChecked,
  disabled = false,
  error,
  inputRef,
  label,
  name,
  onChange,
  readOnly = false,
  value,
}: ToggleProps) => {
  const mergedRefs = mergeRefs(inputRef);

  return (
    <label
      className={classNames(
        css.Toggle,
        error && css.error,
        readOnly && css.readOnly,
        className,
      )}
      data-testid="Toggle"
    >
      <input
        className={css.input}
        checked={checked}
        defaultChecked={defaultChecked}
        disabled={disabled}
        name={name}
        onChange={onChange}
        ref={mergedRefs}
        readOnly={readOnly}
        type="checkbox"
        value={value}
      />
      <div className={css.container}>
        <div className={css.switch}>
          <span className={css.ball} />
        </div>
        {label && <Text className={css.text}>{label}</Text>}
      </div>
    </label>
  );
};

export { Toggle };

export default memo(Toggle);
