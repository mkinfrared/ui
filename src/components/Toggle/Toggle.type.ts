import { CheckboxProps } from "components/Checkbox";

export type ToggleProps = CheckboxProps;
