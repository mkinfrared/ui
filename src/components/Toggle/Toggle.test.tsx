import { render } from "@testing-library/react";
import React from "react";

import { Toggle } from "./Toggle";

describe("<Toggle />", () => {
  const Component = <Toggle />;

  it("should be defined", () => {
    expect(Toggle).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should render a Text component when label is truthy", () => {
    const label = "marklar";
    const { getByTestId } = render(<Toggle label={label} />);
    const labelText = getByTestId("Text");

    expect(labelText).toBeDefined();
  });

  it("should add an error class to label", () => {
    const { getByTestId } = render(<Toggle error />);
    const label = getByTestId("Toggle");

    expect(label.classList).toContain("error");
  });
});
