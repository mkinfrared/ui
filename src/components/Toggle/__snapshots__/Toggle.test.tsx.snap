// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should match the snapshot 1`] = `
<div>
  <label
    class="Toggle"
    data-testid="Toggle"
  >
    <input
      class="input"
      type="checkbox"
      value=""
    />
    <div
      class="container"
    >
      <div
        class="switch"
      >
        <span
          class="ball"
        />
      </div>
    </div>
  </label>
</div>
`;
